src/box/applier.cc

```diff
@@ -369,6 +369,193 @@ applier_connection_init(struct iostream *io, const struct uri *uri,
+applier_run_ballot_triggers(struct applier *applier, bool success)
+{
+	trigger_run(&applier->on_ballot_update, &success);
+}
+
+/**
+ * Perform a IPROTO_VOTE exchange: send a request and decode the response.
+ * Fallback to this method if master isn't avare of "internal.ballot" event.
+ */
+static void
+applier_get_ballot_from_vote(struct iostream *io, struct xrow_header *row,
+			     struct ibuf *ibuf, struct ballot *ballot)
+{
+	RegionGuard guard(&fiber()->gc);
+	xrow_encode_vote(row);
+	coio_write_xrow(io, row);
+	coio_read_xrow(io, ibuf, row);
+	if (row->type == IPROTO_OK) {
+		xrow_decode_ballot_xc(row, ballot);
+		return;
+	}
+	xrow_decode_error(row);
+	struct diag *diag = diag_get();
+	struct error *e = diag_last_error(diag);
+	/*
+	 * Master isn't aware of IPROTO_VOTE request.
+	 * It's OK - we can proceed without it.
+	 */
+	if (box_error_code(e) == ER_UNKNOWN_REQUEST_TYPE) {
+		diag_clear(diag);
+		return;
+	}
+	error_raise(e);
+}
+
+/**
+ * Perform one round of IPROTO_WATCH("internal.ballot") exchange: send the
+ * request / acknowledge that previous notification is accepted, wait for
+ * response.
+ */
+static void
+applier_get_ballot_from_event(struct iostream *io, struct xrow_header *row,
+			      struct ibuf *ibuf, struct ballot *ballot,
+			      int *decoded_fields)
+{
+	RegionGuard guard(&fiber()->gc);
+	xrow_encode_watch_key(row, box_ballot_event_key, IPROTO_WATCH);
+	coio_write_xrow(io, row);
+	coio_read_xrow(io, ibuf, row);
+	if (row->type != IPROTO_EVENT) {
+		xrow_decode_error_xc(row);
+		unreachable();
+	}
+	struct watch_request watch_request;
+	xrow_decode_watch_xc(row, &watch_request);
+	if (watch_request.data == NULL)
+		tnt_raise(ClientError, ER_NO_SUCH_EVENT, box_ballot_event_key);
+	xrow_decode_ballot_event_xc(&watch_request, ballot, decoded_fields);
+}
+
+/** Get applier ballot via the newest of available methods. */
+static void
+applier_get_ballot(struct applier *applier)
+{
+	struct xrow_header row;
+	struct iostream io;
+	struct iostream_ctx io_ctx;
+	struct ibuf ibuf;
+	iostream_ctx_clear(&io_ctx);
+	if (iostream_ctx_create(&io_ctx, IOSTREAM_CLIENT, &applier->uri) != 0)
+		diag_raise();
+	iostream_clear(&io);
+	ibuf_create(&ibuf, &cord()->slabc, 1024);
+	auto guard = make_scoped_guard([&] {
+		if (iostream_is_initialized(&io))
+			iostream_close(&io);
+		iostream_ctx_destroy(&io_ctx);
+		ibuf_destroy(&ibuf);
+	});
+
+	struct greeting greeting;
+
+	applier->addr_len = sizeof(applier->addrstorage);
+	applier_connection_init(&io, &applier->uri, &applier->addr,
+				&applier->addr_len, &io_ctx, &greeting);
+	if (!iproto_features_test(&applier->id.features,
+				  IPROTO_FEATURE_WATCHERS)) {
+		goto try_vote;
+	}
+	try {
+		while (true) {
+			int decoded_fields = 0;
+			applier_get_ballot_from_event(&io, &row, &ibuf,
+						      &applier->ballot,
+						      &decoded_fields);
+			/*
+			 * Some events received right after master start
+			 * might be empty. Ignore those.
+			 */
+			if (decoded_fields > 0)
```

Comment:
I'd rather not.
Having one exchange in `applier_get_ballot_from_event` makes it easier to clear the region.
Now it's enough to have `RegionGuard` in `applier_get_ballot_from_event`.
But if I put the loop inside applier_get_ballot_from_event, I'll have to do both: a region guard in case of an exception, and a region_truncate on each loop iteration. So that in some rare crazy cases applier's memory usage doen't grow indefinitely.